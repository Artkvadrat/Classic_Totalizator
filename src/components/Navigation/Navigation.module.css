.menu {
  display: flex;
  gap: 1em;
  list-style-type: none;
  padding: 1em 0;
  border-bottom: 1px solid black;
}

.menu li:last-child {
  margin-left: auto;
}
