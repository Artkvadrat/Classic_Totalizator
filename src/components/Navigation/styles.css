.ant-menu-horizontal > .ant-menu-submenu:hover,
.ant-menu-horizontal > .ant-menu-item-active,
.ant-menu-horizontal > .ant-menu-submenu-active,
.ant-menu-horizontal > .ant-menu-item-open,
.ant-menu-horizontal > .ant-menu-submenu-open,
.ant-menu-horizontal > .ant-menu-item-selected,
.ant-menu-horizontal > .ant-menu-submenu-selected {
  border-bottom: 2px solid transparent !important;
}

.ant-menu-horizontal a,
.ant-menu-item-selected {
  color: rgba(0, 0, 0, 0.85) !important;
}

.ant-menu-horizontal > .ant-menu-item:hover a {
  color: #1890ff !important;
}

.ant-dropdown-menu-item {
  display: flex;
}
