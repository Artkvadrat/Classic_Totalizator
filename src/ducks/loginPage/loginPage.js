import HTTPService from '../../services/HTTPService/HTTPService';

export const REQUESTED_JWT_TOKEN = 'login_page/requested';
export const RECEIVED_JWT_TOKEN = 'login_page/received';
export const TESTED_JWT_TOKEN = 'login_page/tested';
export const DENIED_JWT_TOKEN = 'login_page/denied';
export const LOGOUT = 'login_page/logout';

const requestedJwtToken = () => ({
  type: REQUESTED_JWT_TOKEN
});

const receivedJwtToken = (data) => ({
  type: RECEIVED_JWT_TOKEN,
  payload: data
});

const testedJwtToken = (data) => ({
  type: TESTED_JWT_TOKEN,
  payload: data
});

const deniedJwtToken = () => ({
  type: DENIED_JWT_TOKEN
});

const logoutAction = () => ({
  type: LOGOUT
});

export const loginUser = ({ email, password }) => (dispatch) => {
  dispatch(requestedJwtToken());

  return HTTPService.request({
    method: 'POST',
    path: '/api/Auth/admin/login',
    body: {
      login: email,
      password
    }
  })
    .then(({ jwtString }) => {
      dispatch(receivedJwtToken(jwtString));
    })
    .catch(() => {
      dispatch(deniedJwtToken());
    });
};

export const testJwtTokenFromLocalStorage = (jwtToken) => (dispatch) => {
  dispatch(requestedJwtToken());

  return HTTPService.request({ path: '/api/Events/outcomes' })
    .then(() => {
      dispatch(testedJwtToken(jwtToken));
    })
    .catch(() => {
      dispatch(deniedJwtToken());
    });
};

export const logout = () => (dispatch) => {
  dispatch(logoutAction());
};

const initialState = {
  isLoading: false,
  isLoggedIn: false
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED_JWT_TOKEN:
      return {
        ...state,
        isLoading: true
      };
    case RECEIVED_JWT_TOKEN:
      window.localStorage.setItem('jwtToken', action.payload);
      return {
        ...state,
        isLoading: false,
        isLoggedIn: true
      };
    case TESTED_JWT_TOKEN:
      return {
        ...state,
        isLoading: false,
        isLoggedIn: true
      };
    case DENIED_JWT_TOKEN:
      window.localStorage.clear();
      return {
        ...state,
        isLoading: false,
        isLoggedIn: false
      };
    case LOGOUT:
      window.localStorage.clear();
      return {
        ...state,
        isLoggedIn: false
      };
    default:
      return state;
  }
};

export default reducer;
