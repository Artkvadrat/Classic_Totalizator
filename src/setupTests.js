import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
