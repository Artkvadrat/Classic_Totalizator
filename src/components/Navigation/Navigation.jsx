import React, { useMemo } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { Menu } from 'antd';

import { logout } from '../../ducks/loginPage/loginPage';

const Navigation = () => {
  const dispatch = useDispatch();

  const logoutHandler = useMemo(
    () => () => {
      dispatch(logout());
    },
    [dispatch]
  );

  return (
    <Menu mode="horizontal">
      <Menu.Item>
        <Link to="/">All Events</Link>
      </Menu.Item>

      <Menu.Item>
        <Link to="/create">Create Event</Link>
      </Menu.Item>

      <Menu.Item>
        <Link to="/users">Users</Link>
      </Menu.Item>

      <Menu.Item style={{ float: 'right' }}>
        <button
          type="button"
          onClick={logoutHandler}
          style={{ backgroundColor: 'inherit', border: 0, cursor: 'pointer' }}
        >
          Logout
        </button>
      </Menu.Item>
    </Menu>
  );
};

export default Navigation;
