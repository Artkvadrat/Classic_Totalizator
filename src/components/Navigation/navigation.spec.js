import React from 'react';
import { shallow } from 'enzyme';
import { Menu } from 'antd';
import { useDispatch } from 'react-redux';
import { logout } from '../../ducks/loginPage/loginPage';

import '../../setupTests';

import Navigation from './Navigation';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn()
}));

jest.mock('../../ducks/loginPage/loginPage', () => ({
  logout: jest.fn()
}));

describe('Navigation component', () => {
  let wrapper;

  const dispatch = jest.fn();
  const mockedLogout = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    logout.mockReturnValue(mockedLogout);

    wrapper = shallow(<Navigation />);
  });

  it('should render without errors', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render 4 menu elements', () => {
    const items = wrapper.find(Menu.Item);

    expect(items.length).toBe(5);
  });

  it('should handle click on button and dispatch logout action', () => {
    const buttonItem = wrapper.find('button');
    buttonItem.simulate('click');

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(logout());
    expect(logout).toHaveBeenCalled();
  });
});
