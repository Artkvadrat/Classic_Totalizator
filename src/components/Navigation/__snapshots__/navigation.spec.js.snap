// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should render without errors 1`] = `
<Menu
  mode="horizontal"
>
  <MenuItem>
    <Link
      to="/"
    >
      All Events
    </Link>
  </MenuItem>
  <MenuItem>
    <Link
      to="/create"
    >
      Create Event
    </Link>
  </MenuItem>
  <MenuItem>
    <Link
      to="/users"
    >
      Users
    </Link>
  </MenuItem>
  <MenuItem
    style={
      Object {
        "float": "right",
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "inherit",
          "border": 0,
          "cursor": "pointer",
        }
      }
      type="button"
    >
      Logout
    </button>
  </MenuItem>
</Menu>
`;
